### PR TITLE
COM-19990: Listing of eZ Platform bundles

### PR DESCRIPTION
**Pull request by kmadejski**
JIRA issue: [https://jira.ez.no/browse/COM-19990](https://jira.ez.no/browse/COM-19990)

## Description
Necessary template and stylesheet files were created. 
Configuration was placed into `config_ezplatform_page.yml`. 
Both, normal and mobile version was created and tested. 

To make fetching Packagist API possible `knplabs/packagist-api` bundle was used and `composer.json` updated also. Gathering API data is handled by `PackagistServiceProvider`.

Solution use `tedivm/stash` for storing data instead of the database. Cache expiration time was set to 7 days.

### CODE REVIEW

File: app/config/views.yml [hunk 1]
```diff
@@ -47,13 +48,30 @@ ezpublish:
...
+                    bundle_list:
+                        controller: app.controller.bundle:showBundlesListAction
+                        params:
+                            query:
+                                query_type: AppBundle:Children
+                                parameters:
+                                    parent_location_id: '%bundles.location_id%'
+                                    limit: '%bundles.cards_limit%'
+                                    sorting_content_type: bundle
+                                    sorting_field: name
```
Comment (SylvainGuittard) [inline app/config/views.yml]: I don't know if you remember, but we discussed the sort order and we decided to use the `DateModified` of the content item.

File: app/config/views.yml [hunk 2]
```diff
@@ -47,13 +48,30 @@ ezpublish:
...
+                                    sorting_order: asc
+                                assign_results_to: cards
+                        template: full/bundle_list.html.twig
+                        match:
+                            Identifier\ContentType: bundle_list
```
Comment (SylvainGuittard) [inline app/config/views.yml]: Add an empty line after.

File: app/config/views.yml [hunk 3]
```diff
@@ -47,13 +48,30 @@ ezpublish:
...
                 line:
                     blog_post:
                         template: line\blog_post.html.twig
                         match:
                             Identifier\ContentType: blog_post
-
+                    bundle:
+                        template: line\bundle_card.html.twig
+                        match:
+                            Identifier\ContentType: bundle
```
Comment (SylvainGuittard) [inline app/config/views.yml]: Add an empty line after

File: app/config/views.yml [hunk 4]
```diff
@@ -47,13 +48,30 @@ ezpublish:
...
+                        controller: app.controller.bundle:showBundlesListAction
+                        params:
+                            query:
+                                query_type: AppBundle:Children
+                                parameters:
+                                    parent_location_id: '%bundles.location_id%'
+                                    limit: '%bundles.cards_limit%'
+                                    sorting_content_type: bundle
+                                    sorting_field: name
+                                    sorting_order: asc
```
Comment (SylvainGuittard) [inline app/config/views.yml]: It should be `desc`

File: src/AppBundle/QueryType/ChildrenQueryType.php [hunk 5]
```diff
@@ -35,6 +35,11 @@ public function getQuery(array $parameters = [])
                 $parameters['sorting_order'] == 'desc' ? Query::SORT_DESC : Query::SORT_ASC
             );
         }
+        else if (isset($parameters['sorting_clause']) &&
```
Comment (SylvainGuittard) [inline src/AppBundle/QueryType/ChildrenQueryType.php]: Can we do something more generic for the `sortClauses` with the `QueryType` and make it compatible with all the cases?

File: web/assets/css/bundle_card.css [hunk 6]
```diff
@@ -0,0 +1,151 @@
+.bundle-card-line {
+    overflow: hidden;
+    margin-top: 20px;
+}
+
+.bundle-card-line-data {
+    background: rgb(241, 241, 241);
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Please change `rgb(241, 241, 241)` to `#f1f1f1`

File: app/Resources/views/line/bundle_card.html.twig [hunk 7]
```diff
@@ -0,0 +1,47 @@
+<div class="bundle-card-line{% if extra.page is defined %} page-{{ extra.page }}{% endif %}">
+    <div class="bundle-card-line-data">
+        <div class="col-sm-8 col-md-9 col-lg-10 bundle-card-line-content">
+            <h4>
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: I think `h2` makes more sense for SEO purposes

File: web/assets/css/bundle_card.css [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    overflow: hidden;
+    border-top-left-radius: 5px;
+    border-top-right-radius: 5px;
+    padding: 20px 5px;
+}
+
+.bundle-card-line-content h4 {
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Change to h2 (ref: https://github.com/ezsystems/ezplatform-com/pull/80/files#diff-205461f0c399ee7cdb4ed7a7eda49f76R4)

File: web/assets/css/bundle_card.css [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    .bundle-card-line-data {
+        padding: 10px 0;
+        border-radius: 5px;
+    }
+
+    .bundle-card-line-author {
+        overflow: hidden;
+    }
+
+    .bundle-card-line-content h4 {
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Change to h2 (ref: https://github.com/ezsystems/ezplatform-com/pull/80/files#diff-205461f0c399ee7cdb4ed7a7eda49f76R4)

File: web/assets/css/bundle_card.css [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    border-top-right-radius: 5px;
+    padding: 20px 5px;
+}
+
+.bundle-card-line-content h4 {
+    font-weight: 700;
+}
+
+.bundle-card-line-author-username {
+    background: rgb(231,231,231);
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Please change `rgb(231, 231, 231)` to `#e7e7e7`

File: web/assets/css/bundle_card.css [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+.bundle-card-line-author-photo {
+    padding: 15px 25px;
+    background: rgb(255,255,255);
+    text-align: center;
+}
+
+.bundle-card-line-last-update {
+    background: rgb(231,231,231);
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Please change `rgb(255, 255, 255)` to `#fff`

File: web/assets/css/bundle_card.css [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    font-weight: 700;
+}
+
+.bundle-card-line-contributors {
+    background: rgb(217, 217, 217);
+    border-bottom-left-radius: 5px;
+    border-bottom-right-radius: 5px;
+    padding: 20px 30px;
+}
+
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: CS: Please remove extra line

File: web/assets/css/bundle_card.css [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    border-bottom-right-radius: 5px;
+}
+
+.bundle-card-line-last-update span {
+    display: block;
+    font-weight: 700;
+}
+
+.bundle-card-line-contributors {
+    background: rgb(217, 217, 217);
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Please change `rgb(217, 217, 217)` to `#d9d9d9`

File: web/assets/css/bundle_card.css [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+.bundle-card-line-contributors span {
+    display: block;
+    float: left;
+    font-size: 18px;
+    padding-top: 10px;
+}
+
+.bundle-card-line-contributor {
+    background: rgb(255,255,255);
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Please change `rgb(255, 255, 255)` to `#fff`

File: web/assets/css/bundle_card.css [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    .bundle-card-line-author {
+        overflow: hidden;
+    }
+
+    .bundle-card-line-content h4 {
+        margin-top: 0;
+    }
+
+    .bundle-card-line-author-container {
+        background: rgb(231,231,231);
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: Please change `rgb(231, 231, 231)` to `#e7e7e7`

File: web/assets/css/bundles_list.css [hunk 16]
```diff
@@ -0,0 +1,73 @@
+.bundles-list {
+    overflow: hidden;
+}
+
+.bundles-list-header {
+    border-top: 1px solid rgb(68,68,68);
+}
+
+    .bundles-list-header .bundles-list-hello-message {
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: CS: Indentation ;)
(and below)

File: web/assets/css/bundles_list.css [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        padding: 50px 0 0 0;
+        font-size: 16px;
+    }
+
+    .bundles-list-header h2 {
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: There is no `h2` in the code.
(before my comment regarding changing `h4` to `h2`)

File: app/Resources/views/full/bundle_list.html.twig [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+                    {% if extraParams.totalCount > items|length %}
+                        <button onclick="loadBundleCards()">Load more</button>
+                    {% endif %}
+                </div>
+            </div>
+        </div>
+    </div>
+    <script>
+        {% autoescape 'js' %}
+        var page = 1,
```
Comment (damianz5) [inline app/Resources/views/full/bundle_list.html.twig]: please wrap the code with 
```
$(document).ready(function() { });   
or
$(function() { });   
```
to be sure that DOM will be ready for jquery selectors

File: web/assets/css/bundles_list.css [hunk 19]
```diff
@@ -0,0 +1,73 @@
+.bundles-list {
+    overflow: hidden;
+}
+
+.bundles-list-header {
+    border-top: 1px solid rgb(68,68,68);
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: Please change `rgb(68,68,68)` to `#444`

File: web/assets/css/bundles_list.css [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    .bundles-list-header h2 {
+        padding: 6px 0;
+        margin: 0;
+        font-weight: 700;
+    }
+
+.bundles-list-content .bundles-list-container {
+    border-top: 1px solid rgb(151, 151, 151);
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: please change `rgb(151, 151, 151)` to `#979797`

File: web/assets/css/bundles_list.css [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+.bundles-list-load-more {
+    width: 100%;
+    overflow: hidden;
+    padding: 20px 0 0 0;
+    text-align: center;
+}
+
+    .bundles-list-load-more button {
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: CS: please remove the indentation
(and below)

File: web/assets/css/bundles_list.css [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+        background: transparent;
+        text-transform: uppercase;
+        padding: 12px 60px;
+        border: 1px solid rgb(51,51,51);
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: Please change `rgb(51, 51, 51)` to `#333`

File: web/assets/css/bundles_list.css [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+        font-size: 16px;
+        font-weight: 700;
+    }
+
+    .bundles-list-load-more button:hover {
+        background: #b1b1b1;
+        color: #fff;
+    }
+
+    .bundles-list-load-progress {
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: Can we define a global spinner for all the website?
ping @damianz5

File: web/assets/css/bundle_card.css [hunk 24]
```diff
@@ -0,0 +1,151 @@
+.bundle-card-line {
```
Comment (SylvainGuittard) [inline web/assets/css/bundle_card.css]: I think the name of the file can be `bundle.css`

File: app/Resources/views/full/bundle_list.html.twig [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+            loadMoreButton = $('.bundles-list-load-more').html();
+
+        function loadBundleCards() {
```
Comment (damianz5) [inline app/Resources/views/full/bundle_list.html.twig]: could be replaced with 
```
$('.bundles-list-load-more button').on('click', function ({

});
```

File: app/Resources/views/line/bundle_card.html.twig [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+        </div>
+    </div>
+    {% if packageDetails.maintainers is defined %}
+    <div class="hidden-xs col-sm-12 bundle-card-line-contributors">
+        <div class="bundle-card-line-contributors-container">
+            <span>By:</span>
+            {% for maintainer in packageDetails.maintainers|slice(0, 4) %}
+            <div class="bundle-card-line-contributor">
+                <div class="bundle-card-line-contributor-photo">
+                    <img src="{{ maintainer.avatarUrl }}&s=40" />
```
Comment (damianz5) [inline app/Resources/views/line/bundle_card.html.twig]: consider changing `&` to `&amp;`

File: app/config/routing.yml [hunk 27]
```diff
@@ -1,4 +1,4 @@
-login:
+xlogin:
```
Comment (damianz5) [inline app/config/routing.yml]: probably a typo

File: src/AppBundle/Controller/BundleController.php [hunk 28]
```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace AppBundle\Controller;
+
+
```
Comment (damianz5) [inline src/AppBundle/Controller/BundleController.php]: remove one new line

File: src/AppBundle/Controller/BundleController.php [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    private $templating;
+
+    /**
+     * @var \eZ\Publish\API\Repository\SearchService
+     */
+    private $searchService;
+
+    /**
+     * @var ChildrenQueryType
```
Comment (damianz5) [inline src/AppBundle/Controller/BundleController.php]: use full namespace

File: src/AppBundle/Controller/BundleController.php [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    private $bundlesListCardsLimit;
+
+    /**
+     * BundleController constructor.
+     * @param EngineInterface $templating
+     * @param SearchService $searchService
+     * @param ChildrenQueryType $childrenQueryType
+     * @param PackagistServiceProviderInterface $packagistServiceProvider
+     * @param $bundlesListLocationId
```
Comment (damianz5) [inline src/AppBundle/Controller/BundleController.php]: add `int` after `@param`, here and at line 51

File: app/Resources/views/line/bundle_card.html.twig (hunk 7 above)
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Done.

File: app/Resources/views/line/bundle_card.html.twig [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+                <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}" title="See on Packagist.org">{{ ez_render_field(content, 'bundle_id') }}</a>
+            </h4>
+            <div class="bundle-card-line-description">
+                {{ ez_render_field(content, 'description') }}
+            </div>
+        </div>
+        <div class="col-sm-4 col-md-3 col-lg-2 bundle-card-line-author">
+            <div class="bundle-card-line-author-container">
+                {% if packageDetails.author[0] is defined %}
+                <div class="bundle-card-line-author-username">{{ packageDetails.author[0].name }}</div>
```
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Done.

File: app/Resources/views/line/bundle_card.html.twig [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+                {% endif %}
+                {% if packageDetails.authorAvatarUrl is defined and packageDetails.author[0].name is defined %}
+                <div class="bundle-card-line-author-photo">
```
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Done.

File: app/Resources/views/line/bundle_card.html.twig [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+                    <img src="{{ packageDetails.authorAvatarUrl }}?s=100" alt="{{ packageDetails.author[0].name }}" />
+                </div>
+                {% endif %}
+                {% if packageDetails.updated is defined %}
+                <div class="bundle-card-line-last-update">Updated: <span>{{ packageDetails.updated|date("d / m / Y") }}</span></div>
```
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Done

File: app/Resources/views/line/bundle_card.html.twig [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+                {% endif %}
+            </div>
+        </div>
+    </div>
+    {% if packageDetails.maintainers is defined %}
+    <div class="hidden-xs col-sm-12 bundle-card-line-contributors">
```
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Done.

File: app/Resources/views/line/bundle_card.html.twig [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+        <div class="bundle-card-line-contributors-container">
+            <span>By:</span>
+            {% for maintainer in packageDetails.maintainers|slice(0, 4) %}
+            <div class="bundle-card-line-contributor">
```
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Done.

File: app/Resources/views/line/bundle_card.html.twig (hunk 26 above)
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Good advice, fixed.

File: app/config/config_ezplatform_page.yml [hunk 36]
```diff
@@ -17,6 +17,12 @@ parameters:
     blog.posts_limit: 10
     # `authors` folder locationId
     blog.authors_container_location_id: 72
+    # `bundles_list` locationId
+    bundles.location_id: 163
+    # 'bundles_list` elements limit on `bundles_list` page
+    bundles.cards_limit: 5
```
Comment (kmadejski) [inline app/config/config_ezplatform_page.yml]: Yes, changed.

File: app/config/routing.yml (hunk 27 above)
Comment (kmadejski) [inline app/config/routing.yml]: Sorry, it was the typo.

File: app/config/views.yml (hunk 1 above)
Comment (kmadejski) [inline app/config/views.yml]: Yes, I remember and I've implemented it in BundleController. This section is not used anymore, so I removed it.

File: app/config/views.yml (hunk 4 above)
Comment (kmadejski) [inline app/config/views.yml]: As above.

File: app/config/views.yml (hunk 2 above)
Comment (kmadejski) [inline app/config/views.yml]: Done.

File: app/config/views.yml (hunk 3 above)
Comment (kmadejski) [inline app/config/views.yml]: Done.

File: src/AppBundle/QueryType/ChildrenQueryType.php (hunk 5 above)
Comment (kmadejski) [inline src/AppBundle/QueryType/ChildrenQueryType.php]: I think that we can implement separate `QueryType` for bundles list.

File: web/assets/css/bundle_card.css (hunk 24 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 6 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 8 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 10 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 11 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 13 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 14 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 9 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundle_card.css (hunk 15 above)
Comment (kmadejski) [inline web/assets/css/bundle_card.css]: Done.

File: web/assets/css/bundles_list.css (hunk 19 above)
Comment (kmadejski) [inline web/assets/css/bundles_list.css]: Done.

File: web/assets/css/bundles_list.css (hunk 16 above)
Comment (kmadejski) [inline web/assets/css/bundles_list.css]: Done.

File: web/assets/css/bundles_list.css (hunk 17 above)
Comment (kmadejski) [inline web/assets/css/bundles_list.css]: Done.

File: web/assets/css/bundles_list.css (hunk 20 above)
Comment (kmadejski) [inline web/assets/css/bundles_list.css]: Done.

File: web/assets/css/bundles_list.css (hunk 21 above)
Comment (kmadejski) [inline web/assets/css/bundles_list.css]: Done.

File: web/assets/css/bundles_list.css (hunk 22 above)
Comment (kmadejski) [inline web/assets/css/bundles_list.css]: Done.

File: app/Resources/views/full/bundle_list.html.twig (hunk 18 above)
Comment (kmadejski) [inline app/Resources/views/full/bundle_list.html.twig]: Done.

File: app/Resources/views/full/bundle_list.html.twig (hunk 25 above)
Comment (kmadejski) [inline app/Resources/views/full/bundle_list.html.twig]: Sure. Fixed.

File: app/Resources/views/line/bundle_card.html.twig (hunk 26 above)
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Good advice, fixed.

File: app/config/routing.yml (hunk 27 above)
Comment (kmadejski) [inline app/config/routing.yml]: Yes, it was the typo, sorry.

File: src/AppBundle/Controller/BundleController.php (hunk 28 above)
Comment (kmadejski) [inline src/AppBundle/Controller/BundleController.php]: Done.

File: src/AppBundle/Controller/BundleController.php (hunk 29 above)
Comment (kmadejski) [inline src/AppBundle/Controller/BundleController.php]: Fixed.

File: src/AppBundle/Controller/BundleController.php (hunk 30 above)
Comment (kmadejski) [inline src/AppBundle/Controller/BundleController.php]: Fixed.

File: src/AppBundle/QueryType/ChildrenQueryType.php (hunk 5 above)
Comment (SylvainGuittard) [inline src/AppBundle/QueryType/ChildrenQueryType.php]: What do you think @damianz5 ?

Comment (damianz5) [inline src/AppBundle/QueryType/ChildrenQueryType.php]: new `QueryType` would be more readable

Comment (SylvainGuittard) [inline src/AppBundle/QueryType/ChildrenQueryType.php]: Let's do it :)

**Pull request merged**